src/hotspot/share/opto/compile.cpp

```diff
@@ -1852,6 +1855,90 @@ void Compile::process_for_post_loop_opts_igvn(PhaseIterGVN& igvn) {
   }
 }
 
+void Compile::record_unstable_if(UnstableIfTrap* trap) {
+  if (AggressiveLivenessForUnstableIf) {
+    _unstable_ifs.append(trap);
+  }
+}
+
+void Compile::invalidate_unstable_if(CallStaticJavaNode* unc) {
+  for (int i = 0; i < _unstable_ifs.length(); ++i) {
+    UnstableIfTrap* trap = _unstable_ifs.at(i);
+    if (trap->uncommon_trap() == unc) {
+      trap->set_bci(-1);
+      break;
+    }
+  }
+}
+
+uint trivial_unstable_ifs          = 0;
+uint unstable_ifs_all              = 0;
+
+void Compile::preprocess_unstable_ifs() {
+  Atomic::add(&unstable_ifs_all, (uint)_unstable_ifs.length());
+
+  for (int i=0; i < _unstable_ifs.length(); i++) {
+    UnstableIfTrap* trap = _unstable_ifs.at(i);
+    if (trap->is_trivial()) {
+      Atomic::inc(&trivial_unstable_ifs);
+    }
+  }
+}
+
+// Re-calculate unstable_if traps with the liveness of next_bci, which points to the unlikely path.
+// It needs to be done after igvn because fold-compares may fuse uncommon_traps and before renumbering.
+void Compile::process_for_unstable_ifs(PhaseIterGVN& igvn) {
+  if (!AggressiveLivenessForUnstableIf)
+    return;
+
+  while (_unstable_ifs.length() > 0) {
+    UnstableIfTrap* trap = _unstable_ifs.pop();
+    CallStaticJavaNode* unc = trap->uncommon_trap();
+    int next_bci = trap->next_bci();
+
+    if (next_bci != -1 && !_dead_node_list.test(unc->_idx)) {
```

Comment:
Did you consider to remove items from `_unstable_ifs` in `Compile::remove_useless_nodes()` to avoid `_dead_node_list` check here?
You would need specialized `remove_useless__unstable_ifs(useful)`